### PR TITLE
Reland "RendererContextSwitch guard flutter's gl context rework. #13812"

### PR DESCRIPTION
**Pull request by cyanglaz**
This PR is an exact re-land of "RendererContextSwitch guard flutter's gl context rework. #13812" + the fix for the image not showing issue which caused the revert. (https://github.com/flutter/flutter/issues/45098)

Added RendererContextSwitches before some other gl operations.

Added a scenario test that makes sure the images are shown. The scenario tests are currently only running on simulator on CI. It needs to be run on an actual device to be sure the gl context is set correctly for image rendering. For now, I have manually tested it locally.
The effort of running the scenario test to device on CI is tracked here: https://github.com/flutter/flutter/issues/43852

Preferably, A framework side golden tests for displaying an image should be added to catch the regression

All the changes in the PR besides "RendererContextSwitch guard flutter's gl context rework. #13812" is inline commented

IMPORTANT: We should only land this when https://github.com/flutter/flutter/issues/45348 is resolved.

### CODE REVIEW

File: shell/common/rasterizer.cc [hunk 1]
```diff
@@ -309,35 +308,58 @@ RasterStatus Rasterizer::DrawToSurface(flutter::LayerTree& layer_tree) {
...
+    return RasterStatus::kFailed;
+  }
+  RasterStatus raster_status = compositor_frame.Raster(layer_tree, false);
+  if (raster_status == RasterStatus::kFailed) {
+    return raster_status;
+  }
+  frame.Submit();
+  return raster_status;
+}
+
```
Comment (cyanglaz) [inline shell/common/rasterizer.cc]: Added switch before `Raster`

File: shell/common/rasterizer.cc [hunk 2]
```diff
@@ -309,35 +308,58 @@ RasterStatus Rasterizer::DrawToSurface(flutter::LayerTree& layer_tree) {
...
     }
 
     FireNextFrameCallbackIfPresent();
 
     if (surface_->GetContext()) {
+      std::unique_ptr<RendererContextSwitchManager::RendererContextSwitch>
+          context_switch = surface_->MakeRenderContextCurrent();
+      if (!context_switch->GetSwitchResult()) {
+        return RasterStatus::kFailed;
+      }
```
Comment (cyanglaz) [inline shell/common/rasterizer.cc]: Added context switch

File: shell/common/rasterizer.cc [hunk 3]
```diff
@@ -309,35 +308,58 @@ RasterStatus Rasterizer::DrawToSurface(flutter::LayerTree& layer_tree) {
...
-static sk_sp<SkData> ScreenshotLayerTreeAsPicture(
+sk_sp<SkData> Rasterizer::ScreenshotLayerTreeAsPicture(
     flutter::LayerTree* tree,
     flutter::CompositorContext& compositor_context) {
   FML_DCHECK(tree != nullptr);
+  std::unique_ptr<RendererContextSwitchManager::RendererContextSwitch>
+      context_switch = surface_->MakeRenderContextCurrent();
+  if (!context_switch->GetSwitchResult()) {
+    return nullptr;
+  }
```
Comment (cyanglaz) [inline shell/common/rasterizer.cc]: Added context switch

File: shell/common/rasterizer.cc [hunk 4]
```diff
@@ -91,6 +91,11 @@ void Rasterizer::NotifyLowMemoryWarning() const {
     FML_DLOG(INFO) << "Rasterizer::PurgeCaches called with no GrContext.";
     return;
   }
+  std::unique_ptr<RendererContextSwitchManager::RendererContextSwitch>
+      context_switch = surface_->MakeRenderContextCurrent();
+  if (!context_switch->GetSwitchResult()) {
+    return;
+  }
```
Comment (cyanglaz) [inline shell/common/rasterizer.cc]: Added context switch

File: shell/common/rasterizer.cc [hunk 5]
```diff
@@ -509,6 +576,11 @@ void Rasterizer::SetResourceCacheMaxBytes(size_t max_bytes, bool from_user) {
   GrContext* context = surface_->GetContext();
   if (context) {
     int max_resources;
+    std::unique_ptr<RendererContextSwitchManager::RendererContextSwitch>
+        context_switch = surface_->MakeRenderContextCurrent();
+    if (!context_switch->GetSwitchResult()) {
+      return;
+    }
```
Comment (cyanglaz) [inline shell/common/rasterizer.cc]: Added context switch

File: shell/common/rasterizer.cc [hunk 6]
```diff
@@ -521,6 +593,11 @@ std::optional<size_t> Rasterizer::GetResourceCacheMaxBytes() const {
   GrContext* context = surface_->GetContext();
   if (context) {
     size_t max_bytes;
+    std::unique_ptr<RendererContextSwitchManager::RendererContextSwitch>
+        context_switch = surface_->MakeRenderContextCurrent();
+    if (!context_switch->GetSwitchResult()) {
+      return std::nullopt;
+    }
```
Comment (cyanglaz) [inline shell/common/rasterizer.cc]: Added context switch

**Comment by chinmaygarde** [comment]
It is a bit hard to parse just the additions so I may have something but LGTM. Can you reword the title to just say something like "Reland "RendererContextSwitch guard flutter's gl context rework.""

**Comment by cyanglaz** [comment]
> It is a bit hard to parse just the additions so I may have something but LGTM. Can you reword the title to just say something like "Reland "RendererContextSwitch guard flutter's gl context rework.""

Done.

**Comment by cyanglaz** [comment]
@chinmaygarde I rebased and it is really to review.
The framework side integration test to catch the image rendering regression is landed.

**Comment by cyanglaz** [comment]
Closing for now. Will open later when I find a reviewer for this.
